### PR TITLE
Task2

### PR DESCRIPTION
**Pull request by nataliatereshchuk**

### CODE REVIEW

File: P2 - JavaScript/src/linearFolding.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        }
+        return previosValue;
+    };
+
+    var unfold = function(callback, initialState){
+        var currentState = initialState;
+        var array = [];
+        while(true) {
+            var result = callback(currentState);
+            if(!result[1]){
```
Comment (mkudosh) [inline P2 - JavaScript/src/linearFolding.js]: Why use array for return value? It's not clear what is meant by 'result[1]'.

More readable solution:

File: P2 - JavaScript/src/linearFolding.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    'use strict';
+    
+    var fold = function(array, callback, initialValue) {
+        if(array.length == 1){
+            return array[0];
+        }
+
+        var previosValue;
+
+        if(arguments.length > 2) {
```
Comment (mkudosh) [inline P2 - JavaScript/src/linearFolding.js]: It would be more readable to check initialValue for existence.

var previousValue = initialValue || array[0].

File: P2 - JavaScript/src/expressions.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        for(var i = 0; i < array.length; i++){
+            callback(array[i], i, array) && result.push(array[i]);
+        }
+        return result;
+    };
+
+    var map = function (array, callback) {
+        var result = [];
+        for(var i = 0; i < array.length; i++) {
+            var element = array[i];
```
Comment (mkudosh) [inline P2 - JavaScript/src/expressions.js]: What is the point of 'element' variable?

File: P2 - JavaScript/src/expressions.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+            if(callback(array[i], i, array)){
+                return array[i];
+            }
+        }
+    };
+
+    var filter = function(array, callback) {
+        var result = [];
+        for(var i = 0; i < array.length; i++){
+            callback(array[i], i, array) && result.push(array[i]);
```
Comment (mkudosh) [inline P2 - JavaScript/src/expressions.js]: Arguable point, but i don't use such shorthand expressions instead of 'if' expression, because readability is low. Too much logic on a single line.

Appropriate use case for such shorthands is: `var v = i || 10;`.

File: P2 - JavaScript/src/memoization.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+functionHelpers.memoization = (function () {
+    'use strict';
+
+    var memoize = function(func) {
+        var memo = {};
+        return function (...args) {
+            if(args in memo) {
+                console.log('args in memo');
```
Comment (mkudosh) [inline P2 - JavaScript/src/memoization.js]: This code is meant for debug purposes. It should be removed.

File: P2 - JavaScript/src/common.js [hunk 6]
```diff
@@ -0,0 +1,2 @@
+var arrayHelpers = {};
+var functionHelpers  = {};
```
Comment (mkudosh) [inline P2 - JavaScript/src/common.js]: This file seems to be redundant. At least it's name doesn't tell anything about it's purpose.

**Comment by mkudosh** [comment]
General points:

1. It's good to learn ES 2015, but it's not possible to use it on every commercial project because of browser requirements. In order to use it, transpiler should be used.

2. Code style. Follow code style guideline more carefully. Don't forget to put semicolons, spaces after keywords (e.g. "if (", "for (").  Use a tool (ESHint, JSJHint, etc) that checks source code.

3. Use typed comparison operators (=== instead of == and so on).

Implementation notes:

- [ ] Lazy. Current implementation is not so lazy as it could be. It only postpones execution, but the result is recalculated every time.

- [ ] Linear fold. Current implementation doesn't fold array with a single element. If i want to sum each element divided by two, i will recieve wrong result for array with a single element.

- [ ] Partial application. Partial application function is not exposed out of it's module, but it's example for a function that sums elements. Logic and examples should be separated.

- [] Memoization. Current implementation doesn't work with objects. Think about using JSON.stringify instead of implicit toString call.

- [] Tests. Tests should include edge cases (e.g. empty arrays). Names should be more decriptive, test suite name and test case name should form a sentence. Example: test suite - 'First', test case - 'even on [3,2,4] should return 2'.

File: P2 - JavaScript/spec/lazyEvaluation_spec.js [hunk 7]
```diff
@@ -0,0 +1,6 @@
+describe("lazy evaluation", function() {
+    it("should return 7", function () {
```
Comment (mkudosh) [inline P2 - JavaScript/spec/lazyEvaluation_spec.js]: This test case should expect functionHelpers.sampleFunctions.sum not to be called before func is called.

**Comment by mkudosh** [comment]
Final notes:

- Be consistent with quotes in code: if you have decided to use single quotes, then use only single quotes everywhere.

- Pay more attention to style guidelines. There are places where spaces are missing after keywords.

- No need to use more callback parameters than it's needed:

`var sum = arrayHelpers.linearFolding.fold(arrayWithEven, function(previosValue, current, i, array) {
    return previosValue + current;
});`

'i' and 'array' parameters are extra here.

- Don't forget to remove unused variables:

` var sumOfRamdomElements = function() {
     var length = 10;
     ...
}`

- Copy-paste carefully to avoid such situations:

`it("Function accepts array [0, 1, 2, 3] and should return [**2**]", function () {
     var arr = [0, 1, 2, 3];
     expect(functionHelpers.derivates.averageOfEvenElements(arr)).toEqual(**1**);
});`

Tests shouldn't be confusing, because they help to document your code and how it works.
